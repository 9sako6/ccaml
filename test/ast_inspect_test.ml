open Ccaml

let delimiter = "=========="

(* with a single return statement with value *)
let program = "int main() {\n    return 29;\n}"
let () = print_endline program
let tokens = Lexer.tokenize program
let ast = Parser.parse tokens
let () = print_string (Ast.inspect ast)
let () = print_endline delimiter

(* with no return statement *)
let program = "int main() {}"
let () = print_endline program
let tokens = Lexer.tokenize program
let ast = Parser.parse tokens
let () = print_string (Ast.inspect ast)
let () = print_endline delimiter

(* with multiple return statements *)
let program = "int main() {\n    return 2;\n    return 100;\n}"
let () = print_endline program
let tokens = Lexer.tokenize program
let ast = Parser.parse tokens
let () = print_string (Ast.inspect ast)
let () = print_endline delimiter

(* with multiple return statements with - *)
let program = "int main() {\n    return -42;\n}"
let () = print_endline program
let tokens = Lexer.tokenize program
let ast = Parser.parse tokens
let () = print_string (Ast.inspect ast)
let () = print_endline delimiter

(* with multiple return statements with ~ *)
let program = "int main() {\n    return ~42;\n}"
let () = print_endline program
let tokens = Lexer.tokenize program
let ast = Parser.parse tokens
let () = print_string (Ast.inspect ast)
let () = print_endline delimiter

(* with multiple return statements with ! *)
let program = "int main() {\n    return !42;\n}"
let () = print_endline program
let tokens = Lexer.tokenize program
let ast = Parser.parse tokens
let () = print_string (Ast.inspect ast)
let () = print_endline delimiter

(* with multiple return statements with !~-~-! *)
let program = "int main() {\n    return !~-~-!42;\n}"
let () = print_endline program
let tokens = Lexer.tokenize program
let ast = Parser.parse tokens
let () = print_string (Ast.inspect ast)
let () = print_endline delimiter
