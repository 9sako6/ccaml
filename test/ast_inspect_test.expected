int main() {
    return 29;
}
Program
 ↳ Function(name: main)
  ↳ Return
   ↳ Const(value: 29)
==========
int main() {}
Program
 ↳ Function(name: main)
==========
int main() {
    return 2;
    return 100;
}
Program
 ↳ Function(name: main)
  ↳ Return
   ↳ Const(value: 2)
  ↳ Return
   ↳ Const(value: 100)
==========
int main() {
    return -42;
}
Program
 ↳ Function(name: main)
  ↳ Return
   ↳ -
    ↳ Const(value: 42)
==========
int main() {
    return ~42;
}
Program
 ↳ Function(name: main)
  ↳ Return
   ↳ ~
    ↳ Const(value: 42)
==========
int main() {
    return !42;
}
Program
 ↳ Function(name: main)
  ↳ Return
   ↳ !
    ↳ Const(value: 42)
==========
int main() {
    return !~-~-!42;
}
Program
 ↳ Function(name: main)
  ↳ Return
   ↳ !
    ↳ ~
     ↳ -
      ↳ ~
       ↳ -
        ↳ !
         ↳ Const(value: 42)
==========
